### PR TITLE
Replace ssh with net=host hostexec pod and kubectl exec in e2e tests where possible

### PR DESCRIPTION
**Pull request by sttts**
- added RunHostCmd to test/e2e/util.go to run a command in a shell in a hostexec container
- ported kubeproxy e2e tests to hostexec
  - speed-up by parallel pod launch
  - speed-up pod delete with zero grace period
  - fix timeouts with blocking udp tests
- ported privilegedPod e2e test to hostexec
- ported NodePort e2e tests to hostexec
- make all of them work with mesos/docker cluster

Prerequisites: https://github.com/kubernetes/kubernetes/pull/17101

_NOTE_: When #17101 is merged, commit 9e9df46 can be removed to use the gcr.io image.

### CODE REVIEW

**Comment by sttts** [comment]
@k8s-bot test this please

**Comment by sttts** [comment]
Shippable fun again.

File: cluster/mesos/docker/docker-compose.yml [hunk 1]
```diff
@@ -85,6 +85,7 @@ apiserver:
     --service-node-port-range=30000-32767
     --cloud-provider=mesos
     --cloud-config=/opt/mesos-cloud.conf
+    --allow-privileged
```
Comment (jdef) [inline cluster/mesos/docker/docker-compose.yml]: is this a leftover from testing, or required to allow remote exec?

File: test/e2e/kubeproxy.go [hunk 2]
```diff
@@ -241,24 +242,18 @@ func (config *KubeProxyTestConfig) dialFromContainer(protocol, containerIP, targ
 func (config *KubeProxyTestConfig) dialFromNode(protocol, targetIP string, targetPort, tries, expectedCount int) {
 	var cmd string
 	if protocol == "udp" {
-		cmd = fmt.Sprintf("echo 'hostName' | nc -w 1 -u %s %d", targetIP, targetPort)
+		cmd = fmt.Sprintf("echo 'hostName' | timeout -t 3 nc -w 1 -u %s %d", targetIP, targetPort)
```
Comment (jdef) [inline test/e2e/kubeproxy.go]: is that really a long enough timeout? worried about introducing flakes

File: test/e2e/kubeproxy.go [hunk 3]
```diff
@@ -457,23 +469,35 @@ func (config *KubeProxyTestConfig) cleanup() {
...
+		runningPods = append(runningPods, rp)
+	}
+
+	return runningPods
 }
 
 func (config *KubeProxyTestConfig) deleteNetProxyPod() {
 	pod := config.endpointPods[0]
-	config.getPodClient().Delete(pod.Name, nil)
+	config.getPodClient().Delete(pod.Name, api.NewDeleteOptions(0))
```
Comment (jdef) [inline test/e2e/kubeproxy.go]: why is the force-delete needed here?

File: test/e2e/service.go [hunk 4]
```diff
@@ -387,17 +387,11 @@ var _ = Describe("Services", func() {
...
-			if err != nil {
-				Expect(err).NotTo(HaveOccurred())
-			}
-			cmd := fmt.Sprintf(`test -n "$(ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN)"`, nodePort)
-			_, _, code, err := SSH(cmd, hosts[0], testContext.Provider)
-			if code != 0 {
-				Failf("expected node port (%d) to be in use", nodePort)
-			}
+		hostExec := LaunchHostExecPod(f.Client, f.Namespace.Name)
+		cmd := fmt.Sprintf(`ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN`, nodePort)
```
Comment (jdef) [inline test/e2e/service.go]: never heard of `ss`; learned something new today :)

File: test/e2e/util.go [hunk 5]
```diff
@@ -1776,7 +1792,12 @@ func NodeSSHHosts(c *client.Client) ([]string, error) {
 			"only found %d external IPs on nodes, but found %d nodes. Nodelist: %v",
 			len(hosts), len(nodelist.Items), nodelist)
 	}
-	return hosts, nil
+
+	sshHosts := make([]string, 0, len(hosts))
+	for _, h := range hosts {
+		sshHosts = append(sshHosts, h+":22")
```
Comment (jdef) [inline test/e2e/util.go]: use net.JoinHostPort here?

File: cluster/mesos/docker/docker-compose.yml (hunk 1 above)
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: This is default in upstream now. There are even e2e tests for this (which are fixed in this PR).

File: test/e2e/kubeproxy.go (hunk 2 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: It should be bigger (or equal) than that of nc. I guess we can decrease it to 1 sec (like the `-w 1`).

File: test/e2e/util.go (hunk 5 above)
Comment (sttts) [inline test/e2e/util.go]: good point

File: test/e2e/kubeproxy.go (hunk 3 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: Why not? It slows down tests without it.

File: cluster/mesos/docker/docker-compose.yml (hunk 1 above)
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: To answer your specific question: it's not needed for remote exec, only for the PrivilegePod e2e tests.

File: test/images/hostexec/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM alpine:3.2
+
+run apk --update add curl netcat-openbsd iproute2 && rm -rf /var/cache/apk/*
+
+CMD [ "sleep", "3600" ]
```
Comment (jdef) [inline test/images/hostexec/Dockerfile]: how did we come up with this time? if the intent is to block until killed can we a bit more explicit? e.g.
```sh
rm -f /fifo && mkfifo /fifo && exec cat </fifo
```

File: test/e2e/service.go [hunk 7]
```diff
@@ -742,17 +736,11 @@ var _ = Describe("Services", func() {
...
-			if err != nil {
-				Expect(err).NotTo(HaveOccurred())
-			}
-			cmd := fmt.Sprintf(`test -n "$(ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN)"`, nodePort)
-			_, _, code, err := SSH(cmd, hosts[0], testContext.Provider)
-			if code == 0 {
-				Failf("expected node port (%d) to not be in use", nodePort)
-			}
+		hostExec := LaunchHostExecPod(f.Client, f.Namespace.Name)
+		cmd := fmt.Sprintf(`! ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN`, nodePort)
```
Comment (jdef) [inline test/e2e/service.go]: don't we want to `grep` before `tail`?

File: test/e2e/service.go (hunk 4 above)
Comment (jdef) [inline test/e2e/service.go]: `grep` before `tail`?

File: test/images/hostexec/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM alpine:3.2
+
+run apk --update add curl netcat-openbsd iproute2 && rm -rf /var/cache/apk/*
```
Comment (jdef) [inline test/images/hostexec/Dockerfile]: why is iproute2 needed?

File: test/e2e/kubeproxy.go [hunk 9]
```diff
@@ -494,11 +518,6 @@ func (config *KubeProxyTestConfig) createPod(pod *api.Pod) *api.Pod {
 	if err != nil {
```
Comment (jdef) [inline test/e2e/kubeproxy.go]: why is this an if statement, vs `Expect(err).NotTo(HaveOccurred())`? just wondering..

File: test/e2e/kubeproxy.go [hunk 10]
```diff
@@ -421,13 +430,16 @@ func (config *KubeProxyTestConfig) setup() {
...
-	hosts, err := NodeSSHHosts(config.f.Client)
-	Expect(err).NotTo(HaveOccurred())
-	config.nodes = make([]string, 0, len(hosts))
-	for _, h := range hosts {
-		config.nodes = append(config.nodes, strings.TrimSuffix(h, ":22"))
+	By("Getting two nodes")
+	nodeList, err := config.f.Client.Nodes().List(labels.Everything(), fields.Everything())
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Failed to get node list: %v", err))
+	config.externalAddrs = NodeAddresses(nodeList, api.NodeExternalIP)
+	if len(config.externalAddrs) < 2 {
```
Comment (jdef) [inline test/e2e/kubeproxy.go]: assuming that this is what allows the test to run on k8s-mesos?

File: test/e2e/kubeproxy.go (hunk 2 above)
Comment (jdef) [inline test/e2e/kubeproxy.go]: i was actually worried about it being long enough, not that it might be too long. if `nc` is timing out after 1s then `timeout -t 3` is probably fine as-is

File: test/e2e/kubeproxy.go (hunk 3 above)
Comment (jdef) [inline test/e2e/kubeproxy.go]: no specific reason as to "why not" -- just wondering why the change was made at all

File: test/e2e/service.go (hunk 4 above)
Comment (sttts) [inline test/e2e/service.go]: the tail only skips a line. Should be fine.

File: test/images/hostexec/Dockerfile (hunk 6 above)
Comment (sttts) [inline test/images/hostexec/Dockerfile]: Well, it's a test image. Hopefully a test does not take an hour. But I will gratefully use your perfect wait command :)

File: test/e2e/service.go (hunk 7 above)
Comment (sttts) [inline test/e2e/service.go]: As above. It's not a `tail -f`. Moreover, I didn't want to touch the actual test.

File: test/images/hostexec/Dockerfile (hunk 8 above)
Comment (sttts) [inline test/images/hostexec/Dockerfile]: For our nice little tool `ss`. Will add a comment.

File: test/e2e/kubeproxy.go (hunk 9 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: Seems like this style is used in the whole file.

File: test/e2e/kubeproxy.go (hunk 10 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: Yeah, ssh only worked on gce, gke, aws.

**Comment by sttts** [comment]
@k8s-bot test this please

**Comment by jdef** [comment]
@ArtfulCoder PTAL

File: test/e2e/util.go [hunk 11]
```diff
@@ -1818,6 +1840,50 @@ func sshCore(cmd, host, provider string, verbose bool) (string, string, int, err
...
+		},
+		ObjectMeta: api.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:            "hostexec",
+					Image:           "sttts/hostexec:1.2",
```
Comment (ArtfulCoder) [inline test/e2e/util.go]: We have to make sure that we change this image to gcr hosted one, before checking in...

Could we split this PR into two where the hostexec image can go in first ?

**Comment by ArtfulCoder** [comment]
I love what you have done here.
Could you split this PR in two, so that we can get hostexec in first ?

File: test/e2e/util.go (hunk 11 above)
Comment (sttts) [inline test/e2e/util.go]: I have prepared a separate PR: https://github.com/kubernetes/kubernetes/pull/17101

**Comment by sttts** [comment]
Rebased after #17101 got merged.

**Comment by sttts** [comment]
@k8s-bot test this

**Comment by sttts** [comment]
@ArtfulCoder rebased and squashed

**Comment by sttts** [comment]
Failing e2e test looks like a flake: "should schedule a pod w/two RW PDs both mounted to one container".

**Comment by sttts** [comment]
@k8s-bot test this

**Comment by sttts** [comment]
Squashed e2e test ports. Rebased to master.

**Comment by sttts** [comment]
Rebased to latest master.

File: test/e2e/kubeproxy.go [hunk 12]
```diff
@@ -288,15 +281,12 @@ func (config *KubeProxyTestConfig) createNetShellPodSpec(podName string) *api.Po
 						{
 							Name:          "udp",
 							ContainerPort: endpointUdpPort,
-						},
-						{
-							Name:          "host",
-							ContainerPort: endpointHttpPort,
-							HostPort:      endpointHostPort,
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: how do we make sure that the pod is created on every host ?

File: cluster/mesos/docker/docker-compose.yml [hunk 13]
```diff
@@ -86,6 +86,7 @@ apiserver:
     --service-node-port-range=30000-32767
```
Comment (ArtfulCoder) [inline cluster/mesos/docker/docker-compose.yml]: Please squash all commits

File: test/e2e/kubeproxy.go (hunk 12 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: Compare the nodeName and line 479.

**Comment by sttts** [comment]
@ArtfulCoder squashed.

**Comment by sttts** [comment]
@k8s-bot test this

>

**Comment by sttts** [comment]
e2e broken again? "Cluster failed to initialize within 300 seconds". Exactly same code as before, only squashed.

**Comment by sttts** [comment]
@k8s-bot test this

**Comment by sttts** [comment]
@ArtfulCoder ptal

**Comment by ArtfulCoder** [comment]
Thank you for the changes!

**Comment by sttts** [comment]
@k8s-bot test this
